x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/ApiKeyServiceTests.java

```diff
@@ -602,6 +610,83 @@ public void testWillGetLookedUpByRealmTypeIfExists() {
         assertEquals("looked_up_by_type", ApiKeyService.getCreatorRealmType(authentication));
     }
 
+    public void testAuthWillTerminateIfGetThreadPoolIsSaturated() throws ExecutionException, InterruptedException {
+        final String apiKey = randomAlphaOfLength(16);
+        final ApiKeyCredentials creds = new ApiKeyCredentials(randomAlphaOfLength(12), new SecureString(apiKey.toCharArray()));
+        writeCredentialsToThreadContext(creds);
+        SecurityMocks.mockGetRequestException(client, new EsRejectedExecutionException("rejected"));
+        ApiKeyService service = createApiKeyService(Settings.EMPTY);
+        final PlainActionFuture<AuthenticationResult> future = new PlainActionFuture<>();
+        service.authenticateWithApiKeyIfPresent(threadPool.getThreadContext(), future);
+        final AuthenticationResult authenticationResult = future.get();
+        assertEquals(AuthenticationResult.Status.TERMINATE, authenticationResult.getStatus());
+        assertThat(authenticationResult.getMessage(), containsString("server is too busy to respond"));
+    }
+
+    public void testAuthWillTerminateIfHashingThreadPoolIsSaturated() throws IOException, ExecutionException, InterruptedException {
+        final String apiKey = randomAlphaOfLength(16);
+        final ApiKeyCredentials creds = new ApiKeyCredentials(randomAlphaOfLength(12), new SecureString(apiKey.toCharArray()));
+        writeCredentialsToThreadContext(creds);
+
+        Hasher hasher = randomFrom(Hasher.PBKDF2, Hasher.BCRYPT4, Hasher.BCRYPT);
+        final char[] hash = hasher.hash(new SecureString(apiKey.toCharArray()));
+        Map<String, Object> sourceMap = buildApiKeySourceDoc(hash);
+        mockSourceDocument(creds.getId(), sourceMap);
+        final ExecutorService mockExecutorService = mock(ExecutorService.class);
+        when(threadPool.executor(ApiKeyService.THREAD_POOL_NAME)).thenReturn(mockExecutorService);
+        Mockito.doThrow(new EsRejectedExecutionException("rejected")).when(mockExecutorService).execute(any(Runnable.class));
```

Comment:
This is not how the `EsThreadPoolExecutor` [handles over-capacity](https://github.com/elastic/elasticsearch/blob/09ff747fe7479f9c695725ec642315cf0343a533/server/src/main/java/org/elasticsearch/common/util/concurrent/EsThreadPoolExecutor.java#L84). The test below as well.

Reply:
Thanks for the great attention to details! 👍  It's now corrected.